File: day3/cpp/main.cpp
#include <cmath>
#include <ctime>
#include <cstdio>
#include <fstream>
#include <map>
#include <sstream>
#include <vector>
#include <string>

bool findSymbol(const std::vector<std::string> &input, int line, int pos, int len)
{
    std::stringstream symbols;
    for (int i = pos - 1; i <= pos + len; i++) {
        symbols << input[line-1][i];
        symbols << input[line+1][i];
    }
    symbols << input[line][pos-1];
    symbols << input[line][pos+len];

    for (const char &c : symbols.str()) {
        if (std::ispunct(c) && c != '.') {
            return true;
        }
    }
    return false;
}

int findGear(const std::vector<std::string> &input, int line, int pos, int len)
{
    int span = input[line].size();

    for (int i = pos - 1; i <= pos + len; i++) {
        if (input[line-1][i] == '*') { return ((line-1) * span + i); }
    }
    for (int i = pos - 1; i <= pos + len; i++) {
        if (input[line+1][i] == '*') { return ((line+1) * span + i); }
    }
    if (input[line][pos-1] == '*') { return (line * span + pos - 1); }
    if (input[line][pos+len] == '*') { return (line * span + pos + len); }

    return -1;
}

int partOne(const std::vector<std::string> &input)
{
    int sum = 0;

    for (int line = 0; line < input.size(); line++) {
        for (int pos = 0; pos < input[line].size(); pos++) {
            if (std::isdigit(input[line][pos])) {
                int num = atoi(&input[line].c_str()[pos]);
                int ndigs = log10(num) + 1;
                if (findSymbol(input, line, pos, ndigs)) {
                    sum += num;
                }
                pos += ndigs;
            }
        }
    }
    return sum;
}

int partTwo(const std::vector<std::string> &input)
{
    std::map<int, std::vector<int>> gears;

    for (int line = 0; line < input.size(); line++) {
        for (int pos = 0; pos < input[line].size(); pos++) {
            if (std::isdigit(input[line][pos])) {
                int num = atoi(&input[line].c_str()[pos]);
                int ndigs = log10(num) + 1;
                int gearid = findGear(input, line, pos, ndigs);
                if (gearid >= 0) {
                    gears[gearid].push_back(num);
                }
                pos += ndigs;
            }
        }
    }

    int sum = 0;
    for (const std::pair<int, std::vector<int>> &key : gears) {
        if (key.second.size() == 2) {
            sum += key.second[0] * key.second[1];
        }
    }
    return sum;
}

std::vector<std::string> readInput(std::stringstream &ss)
{
    std::vector<std::string> lines;
    std::string buffer;
    std::stringstream extraline;

    std::getline(ss, buffer);
    for (int i = 0; i < buffer.size() + 2; i++) {
        extraline << '.';
    }
    lines.push_back(extraline.str());
    lines.push_back('.' + buffer + '.');
    while (std::getline(ss, buffer)) {
        std::stringstream line;
        line << '.' << buffer << '.';
        lines.push_back(line.str());
    }
    lines.push_back(extraline.str());

    return lines;
}

int main()
{
    std::string test = "467..114..\n" \
                        "...*......\n" \
                        "..35..633.\n" \
                        "......#...\n" \
                        "617*......\n" \
                        ".....+.58.\n" \
                        "..592.....\n" \
                        "......755.\n" \
                        "...$.*....\n" \
                        ".664.598..\n";

    struct timespec start, end;
    std::stringstream ss;
    std::ifstream file("..\\input.txt");
    ss << file.rdbuf();
    //ss << test;
    std::vector<std::string> input = readInput(ss);

    timespec_get(&start, TIME_UTC);
    printf("part one = %d\n", partOne(input));
    timespec_get(&end, TIME_UTC);
    printf("time[us]= %d\n", (end.tv_nsec - start.tv_nsec) / 1000);

    timespec_get(&start, TIME_UTC);
    printf("part two = %d\n", partTwo(input));
    timespec_get(&end, TIME_UTC);
    printf("time[us]= %d\n", (end.tv_nsec - start.tv_nsec) / 1000);
    
    return 0;
}

File: day3/cs/main.cs
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using System;

List<string> readInput(string input)
{
    StreamReader sr = new StreamReader(input);
    List<string> lines = new List<string>();
    StringBuilder line = new StringBuilder();
    string extraline;
    string buffer;

    buffer = sr.ReadLine();
    for (int i = 0; i < buffer.Length + 2; i++) {
        line.Append('.');
    }
    extraline = line.ToString();

    lines.Add(extraline);
    lines.Add('.' + buffer + '.');
    line.Clear();
    while ((buffer = sr.ReadLine()) != null) {
        line.Append('.');
        line.Append(buffer);
        line.Append('.');
        lines.Add(line.ToString());
        line.Clear();
    }
    lines.Add(extraline);

    return lines;
}

bool isSym(char c)
{
    return ((c >= '!' && c <= '-') || ( c == '/') || (c >= ':' && c <= '@')); 
}

bool findSymbol(List<string> lines, int line, int pos, int len)
{
    StringBuilder symbols = new StringBuilder();
    for (int i = pos - 1; i <= pos + len; i++) {
        symbols.Append(lines[line - 1][i]);
        symbols.Append(lines[line + 1][i]);
    }
    symbols.Append(lines[line][pos - 1]);
    symbols.Append(lines[line][pos + len]);

    foreach(char c in symbols.ToString()) {
        if (isSym(c)) {
            return true;
        }
    }
    return false;
}

int findGear(List<string> lines, int line, int pos, int len)
{
    int span = lines[line].Length;

    for (int i = pos - 1; i <= pos + len; i++) {
        if (lines[line - 1][i] == '*') { return ((line - 1) * span + i); }
    }
    for (int i = pos - 1; i <= pos + len; i++) {
        if (lines[line + 1][i] == '*') { return ((line + 1) * span + i); }
    }
    if (lines[line][pos - 1] == '*') { return (line * span + pos - 1); }
    if (lines[line][pos + len] == '*') { return (line * span + pos + len); }

    return -1;
}

int partOne(List<string> lines)
{
    int sum = 0;
    for (int line = 0; line < lines.Count; line++) {
        for (int pos = 0; pos < lines[line].Length; pos++) {
            if (Char.IsDigit(lines[line][pos])) {
                string numstr = "" + lines[line][pos];
                int ndigs = 1;
                while (Char.IsDigit(lines[line][++pos])) {
                    numstr += lines[line][pos];
                    ndigs++;
                }
                int num = Int32.Parse(numstr);
                if (findSymbol(lines, line, pos - ndigs, ndigs)) {
                    sum += num;
                }
            }
        }
    }
    return sum;
}

int partTwo(List<string> lines)
{
    Dictionary<int, List<int>> gears = new Dictionary<int, List<int>>();

    for (int line = 0; line < lines.Count; line++) {
        for (int pos = 0; pos < lines[line].Length; pos++) {
            if (Char.IsDigit(lines[line][pos])) {
                string numstr = "" + lines[line][pos];
                int ndigs = 1;
                while (Char.IsDigit(lines[line][++pos])) {
                    numstr += lines[line][pos];
                    ndigs++;
                }
                int num = Int32.Parse(numstr);
                int gearid = findGear(lines, line, pos - ndigs, ndigs);
                if (gearid >= 0) {
                    if (!gears.ContainsKey(gearid)) {
                        gears.Add(gearid, new List<int>(){num});
                    } else {
                        gears[gearid].Add(num);
                    }
                }
            }
        }
    }

    int sum = 0;
    foreach(List<int> val in gears.Values) {
        if (val.Count == 2) {
            sum += val[0] * val[1];
        }
    }
    return sum;
}

List<string> input = readInput(@"..\..\..\..\input.txt");
Stopwatch timer = new Stopwatch();

timer.Start();
int res1 = partOne(input);
timer.Stop();
Console.WriteLine(String.Format("part one = {0}", res1));
Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));

timer.Restart();
int res2 = partTwo(input);
timer.Stop();
Console.WriteLine(string.Format("part two = {0}", res2));
Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));